File: src/main/java/org/docopt/BranchPattern.java
package org.docopt;

import java.util.List;

/**
 * Branch/inner node of a pattern tree.
 */
abstract class BranchPattern extends Pattern {

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result
				+ ((children == null) ? 0 : children.hashCode());
		return result;
	}

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		final BranchPattern other = (BranchPattern) obj;
		if (children == null) {
			return other.children == null;
		}
		else return children.equals(other.children);
	}

	private final List<Pattern> children;

	public BranchPattern(final List<? extends Pattern> children) {
		this.children = Py.INSTANCE.list(children);
	}

	@Override
	public String toString() {
		return String.format("%s(%s)", getClass().getSimpleName(),
				children.isEmpty() ? "" : Py.INSTANCE.join(", ", children));
	}

	@Override
	protected final List<Pattern> flat(final Class<?>... types) {

        if (Py.INSTANCE.in(getClass(), types)) {
			return Py.INSTANCE.list((Pattern) this);
		}

		// >>> return sum([child.flat(*types) for child in self.children], [])
		{
			final List<Pattern> result = Py.INSTANCE.list();

			for (final Pattern child : children) {
				result.addAll(child.flat(types));
			}

			return result;
		}
	}

	public List<Pattern> getChildren() {
		return children;
	}
}

File: src/main/kotlin/org/docopt/BranchPattern.kt
package org.docopt

import kotlin.reflect.KClass

/**
 * Branch/inner node of a pattern tree.
 */
internal abstract class BranchPattern(children: Collection<Pattern?>) : Pattern() {
    override fun hashCode(): Int = (31 * super.hashCode() + children.hashCode())

    override fun equals(other: Any?): Boolean =
        other is BranchPattern && children == other.children

    val children: MutableList<Pattern?> = children.toMutableList()

    override fun toString(): String = String.format(
        "%s(%s)", javaClass.simpleName,
        if (children.isEmpty()) "" else children.joinToString<Any?>(", ")
    )

    override fun flat(vararg types: KClass<*>): List<Pattern> =
        if (types.contains(this::class)) listOf(this)
        else children.flatMap { it!!.flat(*types) }
}